### PR TITLE
Wip/hughsie/proxy mutex

### PR DESCRIPTION
**Pull request by hughsie**
Type of pull request:
- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [X] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

File: libfwupd/fwupd-client.c [hunk 1]
```diff
@@ -535,7 +541,9 @@ fwupd_client_connect_async (FwupdClient *self, GCancellable *cancellable,
 	g_return_if_fail (cancellable == NULL || G_IS_CANCELLABLE (cancellable));
 
 	/* nothing to do */
+	g_mutex_lock (&priv->proxy_mutex);
```
Comment (pwithnall) [inline libfwupd/fwupd-client.c:544]: Holding the lock until the end of `fwupd_client_connect_get_proxy_cb()` means that any other pending calls to `fwupd_client_connect_async()` will block on this line. i.e. The `*_async()` call will become blocking.

I think you probably want to unlock the mutex at end of this function, and lock it again in `fwupd_client_connect_get_bus_cb()` and `fwupd_client_connect_get_proxy_cb()` when you access the shared state in either of them. You’d then have to deal with what might have happened if another thread executed their `fwupd_client_connect_get_bus_cb()`/`fwupd_client_connect_get_proxy_cb()` callback first. i.e. `priv->conn` is unexpectedly not `NULL`, or `priv->proxy` is unexpectedly not `NULL`. In both cases it should be OK to free and drop the newly-returned result from `g_bus_get_finish()` and `g_dbus_proxy_new_finish()` and move on to the next step in the process.

Comment (hughsie) [reply libfwupd/fwupd-client.c:544]: That makes sense; will give it a go, thanks!

Comment (hughsie) [reply libfwupd/fwupd-client.c]: @pwithnall something like this?

File: libfwupd/fwupd-client.c [hunk 2]
```diff
@@ -449,6 +449,13 @@ fwupd_client_connect_get_proxy_cb (GObject *source,
 	g_autoptr(GError) error = NULL;
 	g_autoptr(GVariant) val = NULL;
 	g_autoptr(GVariant) val2 = NULL;
+	g_autoptr(GMutexLocker) locker = g_mutex_locker_new (&priv->proxy_mutex);
+
+	/* another thread did this for us */
+	if (priv->proxy != NULL) {
+		g_task_return_boolean (task, TRUE);
```
Comment (pwithnall) [inline libfwupd/fwupd-client.c:456]: You need to call `g_dbus_proxy_new_finish()` (and then drop the result) or the `g_dbus_proxy_new_async()` call will go unpaired.

Could do something like:
```
g_autoptr(GDBusProxy) proxy = NULL;
g_autoptr(GMutexLocker) locker = NULL;

proxy = g_dbus_proxy_new_finish (res, &error);

locker = g_mutex_locker_new (&priv->proxy_mutex);
if (priv->proxy != NULL)
  {
    g_task_return_boolean (task, TRUE);
    return;
  }
priv->proxy = g_steal_pointer (&proxy);

/* connect signals, etc. */
```

**Review by pwithnall — COMMENTED** [review]
Nice simplification with `g_dbus_proxy_new_for_bus()`

File: libfwupd/fwupd-client.c (hunk 2 above)
Comment (hughsie) [reply libfwupd/fwupd-client.c:456]: > You need to call g_dbus_proxy_new_finish

Doh, of course. I've used your suggestion.

**Comment by pwithnall** [comment]
lgtm now, fwiw

**Comment by hughsie** [comment]
@superm1 don't delete the branch yet pls; someone else is testing with it and it would be confusing to explain to cherry pick from master.

**Pull request merged**